transformations/chinese_butter_fingers_perturbation/transformation.py

```diff
@@ -0,0 +1,80 @@
+import itertools
+import random
+import unidecode as u
+import json
+import os
+
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+def butter_finger(text, prob=0.2, keyboard="pinyin", seed=0, max_outputs=1):
+    random.seed(seed)
+
+    prob_of_typo = prob * 100
+    perturbed_texts = []
+    for _ in itertools.repeat(None, max_outputs):
+        butter_text = ""
+        for chinese_character in text:
+            similar_pinyins = get_characters_with_similar_pinyin(chinese_character, keyboard)
+
+            if random.choice(range(0, 100)) <= prob_of_typo and similar_pinyins != "":
+                new_chinese_character = random.choice(similar_pinyins)
+            else:
+                new_chinese_character = chinese_character
+
+            butter_text += new_chinese_character
+        perturbed_texts.append(butter_text)
+    return perturbed_texts
+
+def get_characters_with_similar_pinyin(chinese_character, keyboard):
+    chinese_character_database = load_chinese_character_data_into_memory()
+
+    pinyin_for_char_to_be_perturbed = ""
+    if keyboard == "pinyin":
+        for char_dict in chinese_character_database:
+            if(chinese_character == char_dict['word']):
+                pinyin_for_char_to_be_perturbed = char_dict['pinyin']
+
+        chars_with_similar_pinyin = ""
+        for char_dict in chinese_character_database:
+            if(u.unidecode(pinyin_for_char_to_be_perturbed) == u.unidecode(char_dict['pinyin'])):
+                chars_with_similar_pinyin += char_dict['word']
+    else:
+        print("Keyboard not supported.")
+
+    return chars_with_similar_pinyin
+
+
+def load_chinese_character_data_into_memory():
```

Comment:
It would be better to call this only once in the `__init__()` function of `ChineseButterFingersPerturbation`, otherwise the file reading will be executed every time when `generate()` is called.

Reply:
Agreed and Addressed.